h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -175,6 +280,116 @@ private static void append(final int[] rightChildren, final int[] leftChildren,
                 discoveredNodes, pointer, true);
     }
 
+    private static ArrayList<Integer> extractInternalIds(TreeProperties properties) {
+        int nodeId = 0;
+        ArrayList<Integer> nodeIds = new ArrayList<>();
```

Comment:
For large trees/forests, having an array list of `Integer` might actually be much more memory intensive than just having `int[]` - the (typically 64bit) pointers are omitted. The price to pay for that in the current state seems to be passing through the children twice. The antidote to that seems to be making this part of the original recursion.

Reply:
fixed so that it returns int[]. I would leave making this part of original recursion for future optimization.